### PR TITLE
Adding modules to pull the top_5 species kraken/bracken results and add the data in the correct format to the IRIDANext JSON file.

### PR DESCRIPTION
**Pull request by kylacochrane**
This PR adds two module (and updates the corresponding sections in the workflow):

1. top5 runs the TOP_5 process that scrapes the resulting Bracken (adjusted for unclassified read) data and pulls the top 5 species from each sample. This process runs a python script developed by Dan Fornika. 
2. nf-core csvtk/concat module to generate a single CSV file containing the top5 data from all samples run through the pipeline, as is the required formatting for the IRIDANext plugin. 

This PR also updates the conf/iridanext.config file to include the top 5 species data into metadata section of the iridanext.output.json file for each sample.

This PR adds assert statements to the pipeline test (main.nf.test) to ensure the metadata from the test data is as expected. 

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] Make sure your code lints (`nf-core lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).

### CODE REVIEW

File: tests/main.nf.test [hunk 1]
```diff
@@ -47,12 +47,30 @@ nextflow_pipeline {
...
+            // output metadata
+            assert iridanext_metadata.SAMPLE2."abundance_1_name" == "Escherichia coli"
+            assert iridanext_metadata.SAMPLE2."abundance_1_ncbi_taxonomy_id" == "562"
+            assert iridanext_metadata.SAMPLE2."abundance_1_num_assigned_reads" == "631"
+            assert iridanext_metadata.SAMPLE2."abundance_1_fraction_total_reads" == "2.52531"
+            assert iridanext_metadata.SAMPLE2."abundance_2_name" == "Bacteroides fragilis"
+            assert iridanext_metadata.SAMPLE2."abundance_2_ncbi_taxonomy_id" == "817"
+            assert iridanext_metadata.SAMPLE2."abundance_2_num_assigned_reads" == "22"
+            assert iridanext_metadata.SAMPLE2."abundance_2_fraction_total_reads" == "0.08805"
+            assert iridanext_metadata.SAMPLE2."abundance_3_name" == "None"
```
Comment (apetkau) [inline tests/main.nf.test:68]: Could you add `assert` statements for `abundance_4` and `abundance_5` results (which should be all missing values)?

**Review by apetkau — CHANGES_REQUESTED** [review]
This is amazing Kyla. Thanks so much for all your work getting this sorted out 😄 . I have a few comments for you below.

File: bin/bracken_top_n_linelist.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+            output_line[taxonomy_id_field] = "None"
+        output_fields.append(taxonomy_id_field)
+
+        num_assigned_reads_field = "abundance_" + num + "_num_assigned_reads"
+        try:
+            output_line[num_assigned_reads_field] = bracken_report_sorted[n][
+                "new_est_reads"
+            ]
+        except IndexError as e:
+            output_line[num_assigned_reads_field] = 0
```
[the remark this answers is not in the document]
Comment (apetkau) [reply bin/bracken_top_n_linelist.py:77]: I have verified that an empty value in a CSV file will be properly encoded as an empty string by the nf-iridanext plugin: https://github.com/phac-nml/nf-iridanext/pull/18

File: modules/local/top5/main.nf [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+This source file is adapted from the taxon-abundance nextflow pipeline developed by
+Dan Fornika as a work of the BC Center for Disease Control Public Health Laboratory,
+which was distributed as a work within the public domain under the
+Apache Software License version 2.0.
+
+This source file has been adapted to work within our pipeline.
+
+Please refer to the README for more information.
+*/
+process TOP_5 {
```
[the remark this answers is not in the document]
Comment (kylacochrane) [reply modules/local/top5/main.nf]: Of course! Love this as a parameter: [2545fad](https://github.com/phac-nml/speciesabundance/pull/12/commits/2545fad15481450c4f2f3ae97f85b459d847689f)

Changed up the TOP_N process, the corresponding workflow (new channel value), added the parameter to the nextflow.config and nextflow_schema.json, and the README 😸

File: bin/bracken_top_n_linelist.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    for n in range(args.top_n):
+        num = str(n + 1)
+
+        name_field = "abundance_" + num + "_name"
+        try:
+            output_line[name_field] = bracken_report_sorted[n]["name"]
+        except IndexError as e:
+            output_line[name_field] = "None"
```
[the remark this answers is not in the document]
Comment (kylacochrane) [reply bin/bracken_top_n_linelist.py]: This works - results in empty CSV cells when data is missing : [dd084f0](https://github.com/phac-nml/speciesabundance/pull/12/commits/dd084f084c580eeddfeb4134b787eb5cf4989819)

File: tests/main.nf.test (hunk 1 above)
Comment (kylacochrane) [reply tests/main.nf.test]: Of course! Added here: [c739693](https://github.com/phac-nml/speciesabundance/pull/12/commits/c739693be421866c2c8a9b6f9a586caa91078eed)

File: tests/main.nf.test [hunk 5]
```diff
@@ -47,12 +47,30 @@ nextflow_pipeline {
...
 
+            // output files
             assert iridanext_samples.SAMPLE2.findAll { it.path == "adjust/SAMPLE2_S_bracken_abundances.csv" }.size() == 1
             assert iridanext_samples.SAMPLE2.findAll { it.path == "krona/SAMPLE2.html" }.size() == 1
             assert iridanext_samples.SAMPLE2.findAll { it.path == "fastp/SAMPLE2.html" }.size() == 1
             assert iridanext_samples.SAMPLE2.findAll { it.path == "fastp/SAMPLE2_R1_trimmed.fastq.gz" }.size() == 1
             assert iridanext_samples.SAMPLE2.findAll { it.path == "fastp/SAMPLE2_R2_trimmed.fastq.gz" }.size() == 1
+
+            // output metadata
+            assert iridanext_metadata.SAMPLE2."abundance_1_name" == "Escherichia coli"
```
[the remark this answers is not in the document]
Comment (kylacochrane) [reply tests/main.nf.test:63]: Sure can 😄 [c739693](https://github.com/phac-nml/speciesabundance/pull/12/commits/c739693be421866c2c8a9b6f9a586caa91078eed)

File: modules/local/mergecsv/main.nf [hunk 6]
```diff
@@ -0,0 +1,28 @@
+process MERGE_CSV {
```
[the remark this answers is not in the document]
Comment (kylacochrane) [reply modules/local/mergecsv/main.nf]: I agree! Module has been switched out for `csvtk_concat` and workflow has been updated: [dc25c28](https://github.com/phac-nml/speciesabundance/pull/12/commits/dc25c283a8ef26bad7a8d55e0faa8a8ee52b9942)

[[1e69ed6](https://github.com/phac-nml/speciesabundance/pull/12/commits/1e69ed695296d8af2b7ae95b3a534661bcb87f34): updates iridanext.config with the module changes]

File: bin/bracken_top_n_linelist.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    output_fields = ["sample_id", "taxonomy_level"]
+    output_line = {
+        "sample_id": args.sample_id,
+        "taxonomy_level": bracken_report_sorted[0]["taxonomy_lvl"],
+    }
+
+    for n in range(args.top_n):
+        num = str(n + 1)
+
+        name_field = "abundance_" + num + "_name"
```
[the remark this answers is not in the document]
Comment (kylacochrane) [reply bin/bracken_top_n_linelist.py]: I have made the changes to the python script here: [cf88aae](https://github.com/phac-nml/speciesabundance/pull/12/commits/cf88aaec9d0e55d92bd9e36602040dd537358cc3)

Let me know what you think 😄

**Review by apetkau — APPROVED** [review]
This is amazing. Thanks so much for all the work you've put into this Kyla 😄

File: bin/bracken_top_n_linelist.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (apetkau) [reply bin/bracken_top_n_linelist.py]: This looks really good. Thank you so much for making the changes.

**Pull request merged**